libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -635,6 +635,122 @@ string YulUtilFunctions::overflowCheckedIntExpFunction(
 	});
 }
 
+string YulUtilFunctions::overflowCheckedIntLiteralExpFunction(
+	RationalNumberType const& _baseType,
+	IntegerType const& _exponentType,
+	IntegerType const& _commonType
+)
+{
+	solAssert(!_exponentType.isSigned(), "");
+	solAssert(_baseType.isNegative() == _commonType.isSigned(), "");
+	solAssert(_commonType.numBits() == 256, "");
+
+	// Calculates the upperbound for exponentiation, that is, calculate `b`, such that
+	// _base**b <= _maxValue and _base**(b + 1) > _maxValue
+	auto findExponentUpperbound = [](bigint const _base, bigint const _maxValue) -> unsigned
+	{
+		unsigned first = 0;
+		unsigned last = 255;
+		unsigned middle;
+
+		while (first < last)
+		{
+			middle = (first + last) / 2;
+
+			if (
+				// The condition on msb is a shortcut that avoids computing large powers in
+				// arbitrary precision.
+				boost::multiprecision::msb(_base) * middle <= boost::multiprecision::msb(_maxValue) &&
+				boost::multiprecision::pow(_base, middle) <= _maxValue
+			)
+			{
+				if (boost::multiprecision::pow(_base, middle + 1) > _maxValue)
+					return middle;
+				else
+					first = middle + 1;
+			}
+			else
+				last = middle;
+		}
+
+		return last;
+	};
+
+	// Converts a bigint number into u256 (negative numbers represented in two's complement form.)
+	// We assume that `_v` fits in 256 bits.
+	auto bigint2u = [&](bigint const& _v) -> u256
+	{
+		if (_v < 0)
+			return s2u(s256(_v));
+		return u256(_v);
+	};
+
+	string functionName = "checked_exp_" + _baseType.richIdentifier() + "_" + _exponentType.identifier();
+	bigint baseValue = _baseType.isNegative() ?
+		u2s(_baseType.literalValue(nullptr)) :
+		_baseType.literalValue(nullptr);
+	unsigned exponentUpperbound;
+
+	// Bases 0, 1 and -1. There is no overflow or underflow. The optimizer can simplify the `exp`
+	// for these cases.
+	if (baseValue == 0 || baseValue == 1 || baseValue == -1)
+		return m_functionCollector.createFunction(functionName, [&]() {
```

Comment:
Actually wouldn't it be better to only call this once at the beginning and pull all the other code into the lambda function?

Reply:
Also, maybe it would be better to use the same template for all cases and just use something like `<?needsCheck>...</needsCheck>` around the `gt`-check.

Reply:
> Also, maybe it would be better to use the same template for all cases and just use something like <?needsCheck>...</needsCheck> around the gt-check.

We cannot call `findExponentUpperBound` for bases `-1, 1` and `0`. So I think it's better to write them seperately.

> Actually wouldn't it be better to only call this once at the beginning and pull all the other code into the lambda function?

What do you mean by that?

Reply:
It can just return `0` for the irregular bases.

All this code will be executed each time you use an exp operator. Only once it reaches `m_functionCollector.createFunction(`, it will not re-execute the generation of the function. Because of that, we usually put `m_functionCollector.createFunction(` as early as possible and do everything else inside the lambda.

Reply:
I moved the code inside the lambda and added a `<needsOverflowCheck>` template to unify the two cases.